File: Assets/Tests/EditMode/AstralBattleEngineTests.cs
using System.Collections.Generic;
using NUnit.Framework;

namespace AstralWilds.Tests
{
    public sealed class AstralBattleEngineTests
    {
        private static List<AstralDemoMember> MakeParty(int count = 2, int hp = 30)
        {
            var party = new List<AstralDemoMember>(count);
            for (int i = 0; i < count; i++)
                party.Add(new AstralDemoMember { id = "member-" + i, displayName = "Member " + i, hp = hp, maxHp = hp });
            return party;
        }

        private static AstralBattleEngine MakeEngine(List<AstralDemoMember> party, int opponentDamage = 6)
        {
            return new AstralBattleEngine(
                party, new[] { 0, 1 },
                "wild-a", "Wild A", "wild-b", "Wild B",
                opponentDamage);
        }

        /// <summary>
        /// Opponents always start at a fixed 30 HP regardless of the party's HP,
        /// so defeating one from a fresh engine takes multiple rounds. Both slots
        /// attack the same opponent slot each round (guarding is irrelevant here
        /// since only opponent HP matters), cycling rounds until it faints.
        /// </summary>
        private static void DefeatOpponent(AstralBattleEngine engine, int opponentSlot)
        {
            for (int round = 0; round < 10 && !engine.GetOpponent(opponentSlot).defeated; round++)
            {
                engine.QueueAttack(0, opponentSlot, out _);
                engine.QueueAttack(1, opponentSlot, out _);
                engine.TryFinishRound();
            }
        }

        [Test]
        public void Constructor_CreatesTwoOpponentsAndActivatesBothSides()
        {
            var party = MakeParty();
            var engine = MakeEngine(party);

            Assert.That(engine.GetOpponent(0).displayName, Is.EqualTo("Wild A"));
            Assert.That(engine.GetOpponent(1).displayName, Is.EqualTo("Wild B"));
            Assert.That(engine.Battle.PlayerSlots[0].Astral, Is.Not.Null);
            Assert.That(engine.Battle.PlayerSlots[1].Astral, Is.Not.Null);
            Assert.That(engine.Battle.OpponentSlots[0].Astral, Is.Not.Null);
            Assert.That(engine.Battle.OpponentSlots[1].Astral, Is.Not.Null);
        }

        [Test]
        public void QueueAttack_AppliesBasicAttackDamageAndMarksActed()
        {
            var engine = MakeEngine(MakeParty());
            var outcome = engine.QueueAttack(0, 0, out int damage);

            Assert.That(outcome, Is.EqualTo(AstralBattleEngine.ActionOutcome.Applied));
            Assert.That(damage, Is.EqualTo(AstralCombatRules.BasicAttackDamage));
            Assert.That(engine.GetOpponent(0).hp, Is.EqualTo(30 - AstralCombatRules.BasicAttackDamage));
            Assert.That(engine.HasActed(0), Is.True);
            Assert.That(engine.HasActed(1), Is.False);
        }

        [Test]
        public void QueueAttack_SameSlotTwiceInARound_IsRejected()
        {
            var engine = MakeEngine(MakeParty());
            engine.QueueAttack(0, 0, out _);
            var second = engine.QueueAttack(0, 1, out int damage);

            Assert.That(second, Is.EqualTo(AstralBattleEngine.ActionOutcome.SlotAlreadyActed));
            Assert.That(damage, Is.Zero);
        }

        [Test]
        public void QueueAttack_AgainstDefeatedTarget_IsInvalid()
        {
            var engine = MakeEngine(MakeParty());
            // Two hits of 12 does not defeat 30 HP; drive it down with repeated rounds instead.
            for (int i = 0; i < 3; i++)
            {
                engine.QueueAttack(0, 0, out _);
                engine.QueueGuard(1);
                engine.TryFinishRound();
            }
            Assert.That(engine.GetOpponent(0).defeated, Is.True);

            var outcome = engine.QueueAttack(1, 0, out int damage);
            Assert.That(outcome, Is.EqualTo(AstralBattleEngine.ActionOutcome.Invalid));
            Assert.That(damage, Is.Zero);
        }

        [Test]
        public void RepeatedAttacks_DefeatOpponentAndClampHpAtZeroAndSyncBattleState()
        {
            var engine = MakeEngine(MakeParty());
            DefeatOpponent(engine, 0);

            Assert.That(engine.GetOpponent(0).hp, Is.Zero, "HP must clamp at zero, never go negative.");
            Assert.That(engine.GetOpponent(0).defeated, Is.True);
            Assert.That(engine.Battle.OpponentParty.Astrals[0].IsDefeated, Is.True);
        }

        [Test]
        public void AllOpponentsDefeated_TrueOnlyWhenBothAreDefeated()
        {
            var engine = MakeEngine(MakeParty());
            Assert.That(engine.AllOpponentsDefeated(), Is.False);

            DefeatOpponent(engine, 0);
            Assert.That(engine.AllOpponentsDefeated(), Is.False, "Only one opponent is down.");

            DefeatOpponent(engine, 1);
            Assert.That(engine.AllOpponentsDefeated(), Is.True);
        }

        [Test]
        public void QueueArcBurst_HitsBothLivingOpponentsForPerTargetDamage()
        {
            var engine = MakeEngine(MakeParty());
            var outcome = engine.QueueArcBurst(0, out int targetsHit);

            Assert.That(outcome, Is.EqualTo(AstralBattleEngine.ActionOutcome.Applied));
            Assert.That(targetsHit, Is.EqualTo(2));
            Assert.That(engine.GetOpponent(0).hp, Is.EqualTo(30 - AstralCombatRules.ArcBurstDamagePerTarget));
            Assert.That(engine.GetOpponent(1).hp, Is.EqualTo(30 - AstralCombatRules.ArcBurstDamagePerTarget));
        }

        [Test]
        public void QueueArcBurst_SkipsAlreadyDefeatedOpponent()
        {
            var engine = MakeEngine(MakeParty());
            DefeatOpponent(engine, 0);

            var outcome = engine.QueueArcBurst(0, out int targetsHit);

            Assert.That(outcome, Is.EqualTo(AstralBattleEngine.ActionOutcome.Applied));
            Assert.That(targetsHit, Is.EqualTo(1));
        }

        [Test]
        public void QueueGuard_MarksGuardedAndActed()
        {
            var engine = MakeEngine(MakeParty());
            var outcome = engine.QueueGuard(0);

            Assert.That(outcome, Is.EqualTo(AstralBattleEngine.ActionOutcome.Applied));
            Assert.That(engine.IsGuarded(0), Is.True);
            Assert.That(engine.HasActed(0), Is.True);
        }

        [Test]
        public void TryFinishRound_WaitsUntilBothLivingActiveSlotsHaveActed()
        {
            var engine = MakeEngine(MakeParty());
            engine.QueueAttack(0, 0, out _);
            var pending = engine.TryFinishRound();

            Assert.That(pending.Resolved, Is.False, "Slot 1 has not acted yet.");
        }

        [Test]
        public void TryFinishRound_AppliesUnguardedCounterattackDamage()
        {
            var party = MakeParty();
            var engine = MakeEngine(party, opponentDamage: 6);
            engine.QueueAttack(0, 0, out _);
            engine.QueueGuard(1);
            engine.TryFinishRound();

            // Slot 0 acted with Attack (not Guard) so its counterattack lands unguarded.
            Assert.That(party[0].hp, Is.EqualTo(30 - 6));
            // Slot 1 guarded, so its counterattack is reduced to ceil(6/3) = 2.
            Assert.That(party[1].hp, Is.EqualTo(30 - 2));
        }

        [Test]
        public void TryFinishRound_ResetsActedAndGuardedForNextRound()
        {
            var engine = MakeEngine(MakeParty());
            engine.QueueAttack(0, 0, out _);
            engine.QueueGuard(1);
            engine.TryFinishRound();

            Assert.That(engine.HasActed(0), Is.False);
            Assert.That(engine.HasActed(1), Is.False);
            Assert.That(engine.IsGuarded(0), Is.False);
            Assert.That(engine.IsGuarded(1), Is.False);
        }

        [Test]
        public void TryFinishRound_RetargetsCounterattackWhenPrimaryTargetAlreadyDefeated()
        {
            var party = MakeParty(2, hp: 10);
            var engine = MakeEngine(party, opponentDamage: 12);

            // Round 1: slot 0 attacks unguarded (takes the full 12 and faints);
            // slot 1 guards (survives at 10 - ceil(12/3) = 6 HP).
            engine.QueueAttack(0, 0, out _);
            engine.QueueGuard(1);
            var round1 = engine.TryFinishRound();
            Assert.That(round1.Resolved, Is.True);
            Assert.That(party[0].defeated, Is.True, "Slot 0 should have fainted from the unguarded counterattack.");
            Assert.That(party[1].hp, Is.EqualTo(10 - 4));

            // Round 2: slot 0 is defeated and cannot act; slot 1 attacks unguarded.
            engine.QueueAttack(1, 0, out _);
            var round2 = engine.TryFinishRound();

            Assert.That(round2.Resolved, Is.True);
            // Opponent 0's counterattack would normally target slot 0, but that
            // slot already fainted, so it must retarget to the still-living
            // slot 1 instead of being silently dropped.
            Assert.That(round2.Summary, Does.Contain("Member 1"));
            Assert.That(round2.Summary, Does.Contain("fainted"));
            Assert.That(party[1].defeated, Is.True);
        }

        [Test]
        public void TryFinishRound_ReportsPlayerDefeatWhenWholePartyFaints()
        {
            var party = MakeParty(2, hp: 1);
            var engine = MakeEngine(party, opponentDamage: 6);
            engine.QueueAttack(0, 0, out _);
            engine.QueueAttack(1, 1, out _);
            var result = engine.TryFinishRound();

            Assert.That(result.Resolved, Is.True);
            Assert.That(result.PlayerDefeat, Is.True);
            Assert.That(party[0].defeated, Is.True);
            Assert.That(party[1].defeated, Is.True);
        }

        [Test]
        public void FindFirstEligibleParty_SkipsDefeatedMembers()
        {
            var party = MakeParty(3);
            party[0].defeated = true;

            Assert.That(AstralBattleEngine.FindFirstEligibleParty(party, 0), Is.EqualTo(1));
            Assert.That(AstralBattleEngine.FindFirstEligibleParty(party, 2), Is.EqualTo(2));
        }

        [Test]
        public void FindFirstEligibleParty_ReturnsNegativeOneWhenNoneEligible()
        {
            var party = MakeParty(2);
            party[0].defeated = true;
            party[1].defeated = true;

            Assert.That(AstralBattleEngine.FindFirstEligibleParty(party, 0), Is.EqualTo(-1));
        }

        [Test]
        public void AllDefeated_TrueOnlyWhenEveryMemberIsDefeated()
        {
            var party = MakeParty(2);
            Assert.That(AstralBattleEngine.AllDefeated(party), Is.False);

            party[0].defeated = true;
            Assert.That(AstralBattleEngine.AllDefeated(party), Is.False);

            party[1].defeated = true;
            Assert.That(AstralBattleEngine.AllDefeated(party), Is.True);
        }
    }
}

File: Astral_Wilds_Unreal/Astral_Wilds/Source/Astral_Wilds/Tests/AstralBattleEngineTests.cpp
// Astral Wilds - automation tests for the Covenant-of-Two round engine
// (UAstralBattleEngine). Exercises the same 2v2 queue/resolve flow the
// Unity prototype's AstralBattleEngine.cs validated (attack, Arc Burst,
// Guard, opponent counterattack, round resolution), cross-checked against
// real verified Play Mode / Play-in-Editor sessions logged in
// Docs/AI/WorkQueue.md so a regression here is caught before it reaches a
// hand-played session - per the roadmap's "verify every change actually
// compiles and runs" discipline. UAstralBattleEngine is a plain UObject
// that owns no actor/world state, so these run as simple automation tests
// with no level or Play-in-Editor session required.
#include "Misc/AutomationTest.h"

#if WITH_AUTOMATION_TESTS

#include "AstralBattleEngine.h"
#include "AstralCombatRules.h"

namespace AstralBattleEngineTests
{
	TArray<FAstralCombatant> MakeParty()
	{
		TArray<FAstralCombatant> Party;

		FAstralCombatant Primary;
		Primary.Id = TEXT("primary");
		Primary.DisplayName = TEXT("Primary");
		Primary.Hp = Primary.MaxHp = 30;
		Party.Add(Primary);

		FAstralCombatant Companion;
		Companion.Id = TEXT("companion");
		Companion.DisplayName = TEXT("Companion");
		Companion.Hp = Companion.MaxHp = 30;
		Party.Add(Companion);

		return Party;
	}
}

IMPLEMENT_SIMPLE_AUTOMATION_TEST(FAstralBattleEngine_QueueAttack, "AstralWilds.Battle.QueueAttack", EAutomationTestFlags::ApplicationContextMask | EAutomationTestFlags::ProductFilter)
bool FAstralBattleEngine_QueueAttack::RunTest(const FString& Parameters)
{
	TArray<FAstralCombatant> Party = AstralBattleEngineTests::MakeParty();
	const int32 ActiveParty[2] = { 0, 1 };

	UAstralBattleEngine* Engine = NewObject<UAstralBattleEngine>();
	Engine->Initialize(&Party, ActiveParty, TEXT("ember"), TEXT("Ember"), TEXT("frost"), TEXT("Frost"), 6);

	int32 DamageDealt = 0;
	const EAstralActionOutcome Outcome = Engine->QueueAttack(0, 0, DamageDealt);
	TestEqual(TEXT("A valid attack is applied"), (uint8)Outcome, (uint8)EAstralActionOutcome::Applied);
	TestEqual(TEXT("A basic attack deals the documented 12 damage"), DamageDealt, UAstralCombatRules::BasicAttackDamage);
	TestEqual(TEXT("Target HP drops by the attack damage"), Engine->GetOpponent(0).Hp, 30 - UAstralCombatRules::BasicAttackDamage);
	TestTrue(TEXT("Acting slot is marked as having acted"), Engine->HasActed(0));

	int32 SecondDamage = 0;
	const EAstralActionOutcome RepeatOutcome = Engine->QueueAttack(0, 1, SecondDamage);
	TestEqual(TEXT("A slot cannot act twice in the same round"), (uint8)RepeatOutcome, (uint8)EAstralActionOutcome::SlotAlreadyActed);
	TestEqual(TEXT("No damage is dealt on a rejected repeat action"), SecondDamage, 0);
	TestEqual(TEXT("The untouched opponent's HP is unaffected"), Engine->GetOpponent(1).Hp, 30);

	return true;
}

IMPLEMENT_SIMPLE_AUTOMATION_TEST(FAstralBattleEngine_QueueAttack_DefeatsOpponent, "AstralWilds.Battle.QueueAttack.DefeatsOpponent", EAutomationTestFlags::ApplicationContextMask | EAutomationTestFlags::ProductFilter)
bool FAstralBattleEngine_QueueAttack_DefeatsOpponent::RunTest(const FString& Parameters)
{
	TArray<FAstralCombatant> Party = AstralBattleEngineTests::MakeParty();
	const int32 ActiveParty[2] = { 0, 1 };

	UAstralBattleEngine* Engine = NewObject<UAstralBattleEngine>();
	Engine->Initialize(&Party, ActiveParty, TEXT("ember"), TEXT("Ember"), TEXT("frost"), TEXT("Frost"), 6);

	// Three basic attacks at 12 damage each brings a 30 HP opponent to 30-12-12-12, clamped to 0.
	int32 Damage = 0;
	Engine->QueueAttack(0, 0, Damage);
	Engine->MarkActed(1);
	Engine->TryFinishRound();

	Engine->QueueAttack(0, 0, Damage);
	Engine->MarkActed(1);
	Engine->TryFinishRound();

	Engine->QueueAttack(0, 0, Damage);

	TestTrue(TEXT("An opponent reduced to zero HP is defeated"), Engine->GetOpponent(0).bDefeated);
	TestEqual(TEXT("Defeated opponent HP clamps at zero"), Engine->GetOpponent(0).Hp, 0);

	int32 DamageAfterDefeat = 0;
	const EAstralActionOutcome OutcomeAfterDefeat = Engine->QueueAttack(1, 0, DamageAfterDefeat);
	TestEqual(TEXT("Attacking an already-defeated opponent is rejected"), (uint8)OutcomeAfterDefeat, (uint8)EAstralActionOutcome::Invalid);

	return true;
}

IMPLEMENT_SIMPLE_AUTOMATION_TEST(FAstralBattleEngine_ArcBurstHitsLivingOpponentsOnly, "AstralWilds.Battle.ArcBurst.HitsLivingOpponentsOnly", EAutomationTestFlags::ApplicationContextMask | EAutomationTestFlags::ProductFilter)
bool FAstralBattleEngine_ArcBurstHitsLivingOpponentsOnly::RunTest(const FString& Parameters)
{
	TArray<FAstralCombatant> Party = AstralBattleEngineTests::MakeParty();
	const int32 ActiveParty[2] = { 0, 1 };

	UAstralBattleEngine* Engine = NewObject<UAstralBattleEngine>();
	Engine->Initialize(&Party, ActiveParty, TEXT("ember"), TEXT("Ember"), TEXT("frost"), TEXT("Frost"), 6);

	int32 FirstTargetsHit = 0;
	Engine->QueueArcBurst(0, FirstTargetsHit);
	TestEqual(TEXT("Arc Burst hits both living opponents"), FirstTargetsHit, 2);
	TestEqual(TEXT("Each opponent takes the documented 8 Arc Burst damage"), Engine->GetOpponent(0).Hp, 30 - UAstralCombatRules::ArcBurstDamagePerTarget);
	TestEqual(TEXT("Each opponent takes the documented 8 Arc Burst damage"), Engine->GetOpponent(1).Hp, 30 - UAstralCombatRules::ArcBurstDamagePerTarget);
	Engine->MarkActed(1);
	Engine->TryFinishRound();

	// Finish off opponent 0 with two more basic attacks across two more rounds so only opponent 1 remains.
	int32 Damage = 0;
	Engine->QueueAttack(0, 0, Damage);
	Engine->MarkActed(1);
	Engine->TryFinishRound();

	Engine->QueueAttack(0, 0, Damage);
	TestTrue(TEXT("Opponent 0 is defeated after enough basic attacks"), Engine->GetOpponent(0).bDefeated);
	Engine->MarkActed(1);
	Engine->TryFinishRound();

	int32 SecondTargetsHit = 0;
	const int32 OpponentOneHpBefore = Engine->GetOpponent(1).Hp;
	Engine->QueueArcBurst(0, SecondTargetsHit);
	TestEqual(TEXT("Arc Burst only counts living opponents once one is defeated"), SecondTargetsHit, 1);
	TestEqual(TEXT("The defeated opponent takes no further damage"), Engine->GetOpponent(0).Hp, 0);
	TestEqual(TEXT("The surviving opponent still takes Arc Burst damage"), Engine->GetOpponent(1).Hp, OpponentOneHpBefore - UAstralCombatRules::ArcBurstDamagePerTarget);

	return true;
}

IMPLEMENT_SIMPLE_AUTOMATION_TEST(FAstralBattleEngine_GuardReducesCounterattack, "AstralWilds.Battle.Guard.ReducesCounterattack", EAutomationTestFlags::ApplicationContextMask | EAutomationTestFlags::ProductFilter)
bool FAstralBattleEngine_GuardReducesCounterattack::RunTest(const FString& Parameters)
{
	TArray<FAstralCombatant> Party = AstralBattleEngineTests::MakeParty();
	const int32 ActiveParty[2] = { 0, 1 };

	UAstralBattleEngine* Engine = NewObject<UAstralBattleEngine>();
	// The 6-damage profile matches the documented Ember Hollow encounter in Docs/AI/WorkQueue.md
	// (unguarded 30->24, guarded 30->28).
	Engine->Initialize(&Party, ActiveParty, TEXT("ember"), TEXT("Ember"), TEXT("frost"), TEXT("Frost"), 6);

	const EAstralActionOutcome GuardOutcome = Engine->QueueGuard(0);
	TestEqual(TEXT("Guard is a valid action"), (uint8)GuardOutcome, (uint8)EAstralActionOutcome::Applied);
	TestTrue(TEXT("Guard marks the slot as guarded"), Engine->IsGuarded(0));
	TestTrue(TEXT("Guard consumes the slot's action for the round"), Engine->HasActed(0));

	Engine->MarkActed(1);
	const FAstralRoundOutcome RoundOutcome = Engine->TryFinishRound();
	TestTrue(TEXT("Round resolves once both slots have acted"), RoundOutcome.bResolved);

	TestEqual(TEXT("Guarded slot takes reduced (ceil(6/3)=2) counterattack damage, matching the documented 30->28 result"), Party[0].Hp, 28);
	TestEqual(TEXT("Unguarded slot takes the full 6 counterattack damage, matching the documented 30->24 result"), Party[1].Hp, 24);
	TestFalse(TEXT("Guard flag clears after the round resolves"), Engine->IsGuarded(0));

	return true;
}

IMPLEMENT_SIMPLE_AUTOMATION_TEST(FAstralBattleEngine_RoundOnlyResolvesOnceBothSlotsAct, "AstralWilds.Battle.TryFinishRound.WaitsForBothSlots", EAutomationTestFlags::ApplicationContextMask | EAutomationTestFlags::ProductFilter)
bool FAstralBattleEngine_RoundOnlyResolvesOnceBothSlotsAct::RunTest(const FString& Parameters)
{
	TArray<FAstralCombatant> Party = AstralBattleEngineTests::MakeParty();
	const int32 ActiveParty[2] = { 0, 1 };

	UAstralBattleEngine* Engine = NewObject<UAstralBattleEngine>();
	Engine->Initialize(&Party, ActiveParty, TEXT("ember"), TEXT("Ember"), TEXT("frost"), TEXT("Frost"), 6);

	int32 Damage = 0;
	Engine->QueueAttack(0, 0, Damage);

	const FAstralRoundOutcome EarlyOutcome = Engine->TryFinishRound();
	TestFalse(TEXT("A round does not resolve while one active slot still hasn't acted"), EarlyOutcome.bResolved);
	TestEqual(TEXT("The un-acted opponent takes no counterattack damage before the round resolves"), Party[1].Hp, 30);

	Engine->MarkActed(1);
	const FAstralRoundOutcome FinalOutcome = Engine->TryFinishRound();
	TestTrue(TEXT("The round resolves once both slots have acted"), FinalOutcome.bResolved);
	TestFalse(TEXT("Neither party slot is defeated by a single 6-damage counterattack"), FinalOutcome.bPlayerDefeat);

	return true;
}

IMPLEMENT_SIMPLE_AUTOMATION_TEST(FAstralBattleEngine_StaticPartyHelpers, "AstralWilds.Battle.StaticPartyHelpers", EAutomationTestFlags::ApplicationContextMask | EAutomationTestFlags::ProductFilter)
bool FAstralBattleEngine_StaticPartyHelpers::RunTest(const FString& Parameters)
{
	TArray<FAstralCombatant> Party = AstralBattleEngineTests::MakeParty();

	TestEqual(TEXT("The first eligible party member from index 0 is index 0"), UAstralBattleEngine::FindFirstEligibleParty(Party, 0), 0);
	TestFalse(TEXT("A party with a living member is not all-defeated"), UAstralBattleEngine::AllDefeated(Party));

	Party[0].bDefeated = true;
	TestEqual(TEXT("Skips a defeated member to find the next eligible one"), UAstralBattleEngine::FindFirstEligibleParty(Party, 0), 1);

	Party[1].bDefeated = true;
	TestEqual(TEXT("Returns -1 when no eligible member remains"), UAstralBattleEngine::FindFirstEligibleParty(Party, 0), -1);
	TestTrue(TEXT("A party with every member defeated is all-defeated"), UAstralBattleEngine::AllDefeated(Party));

	return true;
}

#endif // WITH_AUTOMATION_TESTS
